### PR TITLE
RSDK-2004 Move setup code to slam repo

### PR DESCRIPTION
**Pull request by zaporter-work**
https://viam.atlassian.net/browse/RSDK-2004

This moves some of the setup code from viamrobotics/rdk/services/slam/builtin.go to viamrobotics/slam/config/setup.go

Related RDK PR: https://github.com/viamrobotics/rdk/pull/2004

### CODE REVIEW

File: config/setup.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	dialMaxTimeoutSec int,
+	logger golog.Logger,
+) (pb.SLAMServiceClient, func() error, error) {
+	ctx, span := trace.StartSpan(ctx, "slam::builtIn::setupGRPCConnection")
+	defer span.End()
+	ctx, timeoutCancel := context.WithTimeout(ctx, time.Duration(dialMaxTimeoutSec)*time.Second)
+	defer timeoutCancel()
+	// Increase the gRPC max message size from the default value of 4MB to 32MB, to match the limit that is set in RDK. This is
+	// necessary for transmitting large pointclouds.
+	maxMsgSizeOption := grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(32 * 1024 * 1024))
```
Comment (nicksanford) [inline config/setup.go:45]: Can we move 32 * 1024 * 1024 into a constant with the comment.

File: config/setup_test.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		test.That(t, err, test.ShouldBeNil)
+	})
+}
+
+func TestSetupDirectories(t *testing.T) {
+	logger := golog.NewTestLogger(t)
+	tempDir, err := os.MkdirTemp("", "*")
+	defer os.RemoveAll(tempDir)
+	test.That(t, err, test.ShouldBeNil)
+	SetupDirectories(tempDir, logger)
```
Comment (nicksanford) [inline config/setup_test.go:47]: Please add a failing test for this if possible.

**Review by nicksanford — APPROVED** [review]
LGTM mod adding a failing test

File: config/setup.go [hunk 3]
```diff
@@ -42,7 +42,8 @@ func SetupGRPCConnection(
 	defer timeoutCancel()
 	// Increase the gRPC max message size from the default value of 4MB to 32MB, to match the limit that is set in RDK. This is
 	// necessary for transmitting large pointclouds.
-	maxMsgSizeOption := grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(32 * 1024 * 1024))
+    maxMessageSize := 32 * 1024 * 1024
```
Comment (nicksanford) [inline config/setup.go]: Can we please move this into a module top-level constant along with the comment around the message size?

**Review by nicksanford — CHANGES_REQUESTED** [review]

**Review by nicksanford — APPROVED** [review]
Approved, mod moving grpc resp size into top level.

File: config/setup.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+// SetupDirectories creates the data directory at the specified path along with
+// its data, map, and config subdirectories.
+func SetupDirectories(dataDirectory string, logger golog.Logger) error {
+	for _, directoryName := range [4]string{"", "data", "map", "config"} {
+		directoryPath := filepath.Join(dataDirectory, directoryName)
+		if _, err := os.Stat(directoryPath); err != nil {
```
Comment (nicksanford) [inline config/setup.go:22]: Why was this changed to be any error?

File: config/setup_test.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		test.That(t, errFoo, test.ShouldBeError)
+	})
+	t.Run("Invalid permissions", func(t *testing.T) {
+		tempDir, err := os.MkdirTemp("", "*")
+		test.That(t, err, test.ShouldBeNil)
+		defer os.RemoveAll(tempDir)
+		noPermsDir := tempDir + "/no_permissions"
+		// create a directory in the temp folder
+		// that doesn't have write permissions
+		err = os.Mkdir(noPermsDir, 0o00)
```
Comment (nicksanford) [inline config/setup_test.go:68]: what does `0o00` mean?

**Review by nicksanford — CHANGES_REQUESTED** [review]

**Comment by nicksanford** [comment]
> Approved, mod moving grpc resp size into top level.

Actually, I now see that the behavior changed.

File: config/setup.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+			logger.Warnf("%v directory does not exist", directoryPath)
```
Comment (nicksanford) [inline config/setup.go:23]: Can we please log the error & change the log to not say the directory doesn't exist if it might fail for a different reason.

File: config/setup_test.go (hunk 5 above)
Comment (nicksanford) [reply config/setup_test.go:68]: Its octal, can we please do `0o000`?

File: config/setup.go (hunk 4 above)
Comment (zaporter-work) [reply config/setup.go:22]: This was changed to be any error to prevent a silent failure if the function was unable to stat the directory for other reasons (ex: the directory is not readable)

File: config/setup.go (hunk 6 above)
Comment (zaporter-work) [reply config/setup.go:23]: How do you feel about the following logs: 

```
    logger.go:130: 2023-03-07T23:44:22.970Z	WARN	config/setup.go:24	setup directories: stat /tmp/2898068953/data: no such file or directory
    logger.go:130: 2023-03-07T23:44:22.970Z	WARN	config/setup.go:24	setup directories: stat /tmp/2898068953/map: no such file or directory
    logger.go:130: 2023-03-07T23:44:22.971Z	WARN	config/setup.go:24	setup directories: stat /tmp/2898068953/config: no such file or directory
```


And for invalid permissions:
```
    logger.go:130: 2023-03-07T23:44:22.971Z	WARN	config/setup.go:24	setup directories: stat /tmp/3520687813/no_permissions/data: permission denied
```

File: config/setup.go (hunk 3 above)
Comment (zaporter-work) [reply config/setup.go]: Done. Is this what you were looking for?

File: config/setup.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+// SetupGRPCConnection uses the defined port to create a GRPC client for communicating with the SLAM algorithms.
+func SetupGRPCConnection(
+	ctx context.Context,
+	port string,
+	dialMaxTimeoutSec int,
+	logger golog.Logger,
+) (pb.SLAMServiceClient, func() error, error) {
+	ctx, span := trace.StartSpan(ctx, "slam::builtIn::setupGRPCConnection")
```
Comment (tessavitabile) [inline config/setup.go:44]: setupGRPCConnection should be SetupGRPCConnection

File: config/setup.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+// This increases the message size from 4MB to 32MB to match the RDK.
+// This is necessary for transmitting large pointclouds.
+var grpcMaxMessageSize = 32 * 1024 * 1024
+
+// SetupDirectories creates the data directory at the specified path along with
+// its data, map, and config subdirectories.
+func SetupDirectories(dataDirectory string, logger golog.Logger) error {
```
Comment (tessavitabile) [inline config/setup.go:23]: Should we also set up the rgb and depth directories, or will that be done in the orbslam module?

**Review by tessavitabile — APPROVED** [review]
LGTM with a question

File: config/setup.go (hunk 6 above)
Comment (nicksanford) [reply config/setup.go]: that seems good. Is that output from an unpushed change? I ask b/c I don't see any log on line 24 in the current code in this branch.

File: config/setup_test.go (hunk 5 above)
Comment (nicksanford) [reply config/setup_test.go]: I still don't see 0o000. Is there an unpushed change?

File: config/setup.go (hunk 6 above)
Comment (zaporter-work) [reply config/setup.go]: It is on line 27 now. Same warning format.

File: config/setup_test.go (hunk 5 above)
Comment (zaporter-work) [reply config/setup_test.go]: Nope, it is 0o000 right now. I haven't pushed anything since last night. It even has a comment explaining the format.

Comment (zaporter-work) [reply config/setup_test.go]: I dont know why you're not seeing it... maybe I messed something up. Not sure.

File: config/setup.go (hunk 8 above)
Comment (zaporter-work) [reply config/setup.go:23]: In my opinion that seems best to be left to the orbslam module as it is implementation-specific and adding those might tie the slam repo to single implementations. However, we could alter this function to accept an array of subdirectories to create. Deferring to @kkufieta

**Review by nicksanford — APPROVED** [review]

**Review by nicksanford — APPROVED** [review]

File: config/setup.go (hunk 8 above)
Comment (kkufieta) [reply config/setup.go:23]: I agree with @zaporter-work. The original plan was indeed to have this part in the orbslam repo: https://github.com/viamrobotics/rdk/blob/3089e392730f48fe5b5c7133258f6a26679f0c56/services/slam/builtin/builtin.go#L127-L143

File: config/setup.go (hunk 7 above)
Comment (kkufieta) [reply config/setup.go:44]: [nit]
```suggestion
	ctx, span := trace.StartSpan(ctx, "config::SetupGRPCConnection")
```

File: config/setup_test.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+		test.That(t, err, test.ShouldBeNil)
+		defer os.RemoveAll(tempDir)
+		noPermsDir := tempDir + "/no_permissions"
+		// create a directory in the temp folder
+		// that doesn't have write permissions
+		// in rwx format: --- --- ---
+		err = os.Mkdir(noPermsDir, 0o000)
+		test.That(t, err, test.ShouldBeNil)
+		err = SetupDirectories(noPermsDir, logger)
+		test.That(t, err, test.ShouldBeError)
```
Comment (kkufieta) [inline config/setup_test.go:72]: [opt/nit/q] Why not check the error message?
```suggestion
		test.That(t, err, test.ShouldBeError)
		test.That(t, fmt.Sprint(err), test.ShouldContainSubstring, "issue creating directory at ")
```

**Review by kkufieta — APPROVED** [review]
LGTM mod nits

File: config/setup_test.go (hunk 9 above)
Comment (zaporter-work) [reply config/setup_test.go]: Agreed. Good ideas for both suggestions.
